### PR TITLE
[StatusBar] Allow status bars to be repainted if they are not covered

### PR DESCRIPTION
**Pull request by zwim**
Currently, if the frontlight is dimmed, the header and footer don't get updated every minute.

This PR fixes this, so that the header and footer show the right time, even if fully dimmed.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9768)

### CODE REVIEW

**Comment by poire-z** [comment]
> if the frontlight is dimmed, the header and footer don't get updated every minute

"is dimmed" ? Or "is dimming" ? I don't really when and how long your TrapWidget gets being on top.
Anyway, do header/footer get updated when reading a dict result (= a DictQuickLookup widget is at top) ?
It feels instead of checking if top widget is ReaderUI or a few other selected ones, we should check if the top widget "covers full screen" (they already got a flag for it).
Anyway, @NiLuJe knows more about all that than I do :)

**Comment by zwim** [comment]
I really mean is dimmed. The TrapWidget is the top level widget, when fully dimmed (we need that to swallow the waking tap).

**Comment by NiLuJe** [comment]
I would *possibly* limit that to the AutoDim TrapWidget, by making sure you pass a distinct name to it at construction, instead of catching *every* TrapWidget?

**Comment by NiLuJe** [comment]
@poire-z's idea of !widget.covers_fullscreen might have some merit now that the whitelist has grown, but I'm not on my dev box to check right now.

(It certainly made sense to limit to ReaderUI when that was the only thing being tested, but now... it maybe warrants a smarter check :)).

**Comment by poire-z** [comment]
> I really mean is dimmed. 
> I would possibly limit that to the AutoDim TrapWidget,

Oh, ok, that plugin that dim the light after some time of inactivity (lack of context with a missing "AutoDim" mention in the title and the first post :)
Well, does it then matter if the time is not updated if 1) lact of activity 2) the light is dimmed and the screen and time hard to read :) ?

**Comment by zwim** [comment]
> 1) lact of activity 2) the light is dimmed and the screen and time hard to read :) ?

ad 1) Maybe the reader lies visible on the table (dimmed to 20%), then the not changing time is irritating.
Ad 2) what if it lies just there, someone turns on the light and sees the wrong time.

All in all it bothers me :) and the fix is not an endless storie.

**Comment by poire-z** [comment]
> the fix is not an endless storie

You bet ? :)

> I would possibly limit that to the AutoDim TrapWidget, by making sure you pass a distinct name to it at construction, instead of catching every TrapWidget?

> idea of !widget.covers_fullscreen might have some merit now that the whitelist has grown

Or may be, instead of naming things, or hijacking covers_fullscreen meaning:
let's just add another property to a widget, similar to how/when `covers_fullscreen` is added: `allow_auto_refreshes_underneath` (or a better name you will come up with :).
We can then set it on ReaderUI (so we don't even have to name ReaderUI), and some TrapWidget but not on others (ie. when saving Wikipedia as EPUBs or batch book metadata updates and there's a middle InfoMessage that is updating quickly, a screen refresh for a small rectangle at bottom may coalesce into a full screen flashing refresh).
Then, we should come up with a generic check/rules at the places needed:
walk from top widget down until a covers_fullscreen widget is met, and expect all widgets met to have this flag. If one doesn't have it, don't autorefresh. (Or something along this idea :)

**Comment by zwim** [comment]
> You bet ? :)

Not any more, but I want to try a commit in your proposed spirit the next days. 
If that proposal seems to be easy going I will continue, else I close this PR (and add this feature to my pers. Koreader )

**Comment by poire-z** [comment]
I think it should be the spirit of how to handle that, but it needs UIManager's manager aka @NiLuJe to confirm or think it a bit further than I did :)

**Comment by zwim** [comment]
@poire-z No worrry you are my master of desaster.





@NiLuJe and you, too.


;-)


For my personal issue, I have a solution. For other users we find one. No hurrrry

**Comment by NiLuJe** [comment]
I'm not sure which is the lesser evil between just piling on widgets in the existing check, or coming up with a brand new widget flag that will only ever be used by this stuff.

I'd tend to prefer the former, so, if you can't find something that works with existing widget flags, I'm probably fine with this as-is, I guess?

**Comment by poire-z** [comment]
> a brand new widget flag that will only ever be used by this stuff

I asked @zwim: don't you mind the time not being auto refreshed when you're reading a dict quick lookup or a wikipedia article?

> which is the lesser evil between just piling on more widgets on a case-by-case basis in the existing check
> maybe just with a dedicated name for the AutoDim widget 

Well, I feel uneasy having many names (even more if these are names from plugins) in a generic UIManager.
Moreover, with flags, we could propagate/cancel the state as widgets are added (haven't though about the logic), so it could be a single check on the top widget, instead of checking the stack (or checking for multiple names).

**Comment by NiLuJe** [comment]
> Well, I feel uneasy having many names (even more if these are names from plugins) in a generic UIManager.

But it isn't :?

The mess is firmly located inside the footer/header aut-refresh codepaths?

**Comment by poire-z** [comment]
> The mess is firmly located inside the footer/header auto-refresh codepaths?

Oh, right :) So, let's let UIManager alone.
Still, having plugin names in core stuff (ReaderFooter) feels bad.
Instead of setting a name to the AutoDim TrapWidget instance, just set a flag "allow_autorefresh" or something like that, and check for it instead of name (as if no name, we get the widget instance from getTopWidget() - it's a bit akward to get either a name or a table, and may crash if we get a name string...)
`if top_wg == "ReaderUI" or (top_wg and type(top_wg) ~= "string" and top_wg.allow_autorefresh) then`

**Comment by NiLuJe** [comment]
Okay, I would see that working (and maybe just write a dedicated variant of of getToWidget(like logic in UIManager that just checks this flag to avoid the type nastiness?)

**Comment by zwim** [comment]
Works well, but not ready right now, as I want to 
1. remove `getSecondTopmostWidget` anymore,
2. remove all added `widget.name` (except `ReaderUI`) and
3. check if the undesirable check for `VirtualKeyboard` can be dropped anywhere.

What I have tested with both status bars enabled:
- [x] Just reading: update
- [x] Menu open: update (but now the menu shows the wrong time, footer OK)
- [x] Dictionary lookup: update, no update if virtkeyboard shown
- [x] ReaderHighlight buttonBox: update 
- [x] Keyboard open: no update 
- [x] Keyboard open and language select dialog open: no update 
- [x] Wikipedia lookup: update  
- [x] Menu open: update
- [x] Menu open and SpinWidget open: update
- [x] Movable dialog not covering footer: update
- [x] Movable dialog covering footer: no update   
- [x] **AutoDim active or not for all of the above cases**


Not tested: Only one status bar enabled, none of them enabled

**Comment by poire-z** [comment]
There is a big frontend/profiler.lua in the middle.

**Comment by zwim** [comment]
> There is a big frontend/profiler.lua in the middle.

Thanks, that will be dropped in any case.

File: frontend/apps/reader/modules/readerfooter.lua [hunk 1]
```diff
@@ -771,6 +771,60 @@ function ReaderFooter:unscheduleFooterAutoRefresh()
...
+--            return false
+            elseif widget.covers_fullscreen or widget.covers_footer == true then
+            -- (e.g. the virtual keyboard sets widget_covers_footer == true)
+            return false
+        elseif widget.ignore ~= "height" and widget.covers_footer ~= false then
+            -- widget.ignore == "height" is set in menu, widget.covers_footer is an optional hint for this here,
+            -- sample usage can be found in the AutoDim plugin.
+
+            -- This is just a simple check if any of the topmost widgets paint into the statusbar's y position.
+            -- Better would be to check if any topmost widget rectancle intersects with the footer content.
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerfooter.lua:807]: `Gem` has an `intersect` method, FWIW.

(But... see the next comment ;)).

File: frontend/apps/reader/modules/readerfooter.lua [hunk 2]
```diff
@@ -771,6 +771,60 @@ function ReaderFooter:unscheduleFooterAutoRefresh()
...
+            -- @todo: xxx
+            local widget_bottom_coordinate = 0 -- a widget may have no dimension (Hello TrapWidget)
+            if widget.movable then
+                widget_bottom_coordinate = widget.movable.dimen.y + widget.movable.dimen.h
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerfooter.lua:811]: `dimen` is... weird.

You have *zero* guarantee that it:

* exists at all
* is actually a Geom and not just a plain table with possibly only a few of the four keys set
* is actually meaningful (a lot of stuff just chucks the Screen dimensions at it).
* is actually up-to-date (a lot of widgets only compute it during `paintTo`).

TL;DR: I'd be *extremely* wary of using `dimen` in any way, shape or form outside of UIManager's `_repaint` method.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua:811]: Case in point: ReaderFooter itself jumps through lot of shitty hoops to compute its own height *before* drawing...

https://github.com/koreader/koreader/blob/a23c30f862aea0cd553fa7491edfc68224e931be/frontend/apps/reader/modules/readerfooter.lua#L2196-L2223

Comment (zwim) [reply frontend/apps/reader/modules/readerfooter.lua:811]: Thanks for your elaborations. In that case I will stop the work on this PR (before it get never ending :) )
Feel free to close it.

**Comment by zwim** [comment]
OK I have dropped the whole dimen stuff.  Drawbacks I see now are, 
* that the contents of a  moveable container moved to the bottom gets overwritten by the footer
* KeyValue widget gets overwritten at the bottom.

I want to look another hour into it this evening, before I decide to close it or make it ready.

File: frontend/apps/reader/modules/readermenu.lua [hunk 3]
```diff
@@ -443,6 +443,7 @@ function ReaderMenu:onShowMenu(tab_index)
     menu_container[1] = main_menu
     -- maintain a reference to menu_container
     self.menu_container = menu_container
+    menu_container.name = "MenuContainer"
```
Comment (poire-z) [inline frontend/apps/reader/modules/readermenu.lua:446]: You added a few `name =`, which seem to not be used anywhere ?

File: frontend/apps/reader/modules/readerfooter.lua [hunk 4]
```diff
@@ -771,6 +771,60 @@ function ReaderFooter:unscheduleFooterAutoRefresh()
     UIManager:unschedule(self.autoRefreshFooter)
 end
 
+function ReaderFooter:toBeRepainted() -- xxx a better name here
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:774]: `:shouldRepaint()` ? (or `:shouldBeRepainted()`)

**Comment by NiLuJe** [comment]
> https://github.com/koreader/koreader/blob/master/plugins/vocabbuilder.koplugin/main.lua#L67-L70

We've discussed this particular interaction before, in the current context VocabBuilder can appear, it's not going to happen. But we're aware of it ;).

**Comment by weijiuqiao** [comment]
I’m not by a computer now, you can go ahead and replace it, if the 2 functions serve the same purpose.
Btw we can still use .tweak_buttons_func on a string variable. Checked with NiLuJe and the result is nil. Very unintuitive :)

**Comment by NiLuJe** [comment]
A couple of random notes:

* Rename to `getNthTopWidget`?
* I would keep the `VirtualKeyboard` name. It doesn't hurt, and will prevent someone else from going mad when they realize you can get multiple VKs stacked on top of each other (IIRC, InputContainer + TextEditor is one fine way of doing that).
* I would confirm that no longer skipping over VKs is not an issue in VocabBuilder's use-case ;).

**Comment by poire-z** [comment]
Looks fine to me, but last word is @NiLuJe 's.

**Comment by NiLuJe** [comment]
(Will give it a proper look in a few hours ;)).

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 2 of 16 files at r2, 1 of 1 files at r3, 7 of 12 files at r4, 7 of 8 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9768)* status: all files reviewed, 9 unresolved discussions (waiting on @poire-z and @zwim)

___
*[`frontend/apps/reader/modules/readerfooter.lua` line 784 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5pqAOB1eXUmN410zO:-NH5pqAOB1eXUmN410zP:b-7xi6zk) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L784)):*
> ```Lua
>             return false
>         end
>         -- Here we land if the widget is not "ReaderUI", wieget.cover_footer ~= nil and widget.ignore == "height"
> ```

Tail end of the comment no longer applies ;).

You can probably get rid of it entirely, now that the logic is much simpler, actually.
___
*[`frontend/apps/reader/modules/readerfooter.lua` line 799 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5q8a4BZsgE0_8BrB4:-NH5q8a4BZsgE0_8BrB5:b-476ji8) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L799)):*
> ```Lua
>             -- be fine refreshing it if the widget is not covering it, but this is hard to
>             -- guess from here.)
>             self:onUpdateFooter(self:shouldBeRepainted())
> ```

This got rid of the `self.view.footer_visible` (i.e., this should only pass true if *both* are true).
___
*[`frontend/apps/reader/modules/readerfooter.lua` line 2477 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qKJq3tw0yrtOvqSz:-NH5qKJq3tw0yrtOvqT-:b-89a4l7) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L2477)):*
> ```Lua
> function ReaderFooter:maybeUpdateFooter()
>     -- ...so we need to avoid stomping over unsuspecting widgets (usually, ScreenSaver).
>     self:onUpdateFooter(self:shouldBeRepainted())
> ```

Ditto
___
*[`frontend/apps/reader/modules/readerfooter.lua` line 2480 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qMdlDdv0eXmYisW4:-NH5qMdlDdv0eXmYisW5:bxs3jvi) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L2480)):*
> ```Lua
> end
> 
> -- is the same as mybeUpdatefooter
> ```

nit: typo
___
*[`frontend/apps/reader/modules/readerfooter.lua` line 2482 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qNLtD6VgNrlsxIsB:-NH5qNLtD6VgNrlsxIsC:b-89a4l7) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L2482)):*
> ```Lua
> -- is the same as mybeUpdatefooter
> function ReaderFooter:onFrontlightStateChanged()
>     self:onUpdateFooter(self:shouldBeRepainted())
> ```

Ditto
___
*[`frontend/apps/reader/modules/readerstatus.lua` line 58 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qXHv0P1s6mM3Fd_-:-NH5qXHv0P1s6mM3Fd_0:b-o7ppy0) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerstatus.lua#L58)):*
> ```Lua
> 
>     local top_widget = UIManager:getNthTopWidget()
>     if (settings == "pop-up" or settings == nil) and top_widget and top_widget.name ~= "end_document" then
> ```

I'm not entirely sure we need the `nil` guard, as UIManager exits if the window stack is empty, but, why not ;).

(I'd rather handle it the other way around, though: `local top_widget = UIManager:getNthTopWidget() or {}`).

EDIT: While a guard might make sense when we check for n > 1, here, it definitely doesn't, so I'd just get rid of it, actually ;).

**Review by zwim — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9768)* status: all files reviewed, 9 unresolved discussions (waiting on @NiLuJe and @poire-z)

___
*[`frontend/apps/reader/modules/readerstatus.lua` line 58 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qXHv0P1s6mM3Fd_-:-NH5y_NaCgcT3EkRgDzp:b-fyre7w) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerstatus.lua#L58)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

I'm not entirely sure we need the `nil` guard, as UIManager exits if the window stack is empty, but, why not ;).

(I'd rather handle it the other way around, though: `local top_widget = UIManager:getNthTopWidget() or {}`).

EDIT: While a guard might make sense when we check for n > 1, here, it definitely doesn't, so I'd just get rid of it, actually ;).
</blockquote></details>

removing the nil guard will lead to a failing readerrolling_spec
```
[ RUN      ] ./spec/front/unit/readerrolling_spec.lua @ 60: Readerrolling module test in portrait screen mode should emit EndOfBook event at the end of sample epub
frontend/apps/reader/modules/readerstatus.lua:58: attempt to index local 'top_widget' (a nil value)

stack traceback:
	frontend/apps/reader/modules/readerstatus.lua:58: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:82: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:100: in function 'handleEvent'
	frontend/apps/reader/modules/readerrolling.lua:867: in function 'onGotoViewRel'
	./spec/front/unit/readerrolling_spec.lua:74: in function <./spec/front/unit/readerrolling_spec.lua:60>

[  ERROR   ] ./spec/front/unit/readerrolling_spec.lua @ 60: Readerrolling module test in portrait screen mode should emit EndOfBook event at the end of sample epub (2.62 ms)
```
-> local top_widget = UIManager:getNthTopWidget() or {} will do for now. Before merging I want to fix readerrolling_spec.lua, and remove that ugly construct.

**Comment by zwim** [comment]
Ready from my side.

**Review by NiLuJe — APPROVED** [review]
> Reviewed 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9768)* status: all files reviewed, 3 unresolved discussions (waiting on @poire-z and @zwim)

___
*[`frontend/apps/reader/modules/readerfooter.lua` line 2480 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qMdlDdv0eXmYisW4:-NHB1yqF3rHc-9Mudhpw:bm25tif) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L2480)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

nit: typo
</blockquote></details>

Still there ;).

**Comment by zwim** [comment]
> > Reviewed 4 of 4 files at r8, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/koreader/koreader/9768)_ status: all files reviewed, 3 unresolved discussions (waiting on @poire-z and @zwim)
> 
> _[`frontend/apps/reader/modules/readerfooter.lua` line 2480 at r7](https://reviewable.io/reviews/koreader/koreader/9768#-NH5qMdlDdv0eXmYisW4:-NHB1yqF3rHc-9Mudhpw:bm25tif) ([raw file](https://github.com/koreader/koreader/blob/b90d35b0b4de3f3fe83907d1f8e8a36f6fa08cdf/frontend/apps/reader/modules/readerfooter.lua#L2480)):_
> Previously, NiLuJe (NiLuJe) wrote…
> 
> Still there ;).

Oh, there were two typos in one word, that's very unlikely (m**a**ybeUpdate**F**ooter).

**Review by NiLuJe — APPROVED** [review]

File: plugins/autodim.koplugin/main.lua [hunk 5]
```diff
@@ -344,7 +324,7 @@ function AutoDim:ramp_task()
         self.ramp_event_countdown = self.ramp_event_countdown - 1
         if self.ramp_event_countdown <= 0 then
             -- Update footer on every self.ramp_event_countdown call
-            self:updateFooter()
+            UIManager:broadcastEvent("UpdateFooter")
```
Comment (poire-z) [inline plugins/autodim.koplugin/main.lua:327]: Just stumble on this line while doing `rgrep broadcast.*ooter frontend plugins`.
I don't think it works if we provide just a string. At least, it's the single bit of code that does that.
The correct way is as it was in the remove chunk above:
`UIManager:broadcastEvent(Event:new("UpdateFooter"...)`.

Comment (NiLuJe) [reply plugins/autodim.koplugin/main.lua:327]: Nice catch ;).

In fact, the only thing that prevents it from crashing is the fact that strings are a metatable, so you *can* try to access a field in a string (yay, Lua), but that's an implementation detail ;o).

On the upside, I think not doing anything is the right call: refreshes would race with the frontlight changes and ruin the effect. Only updating on end makes the most sense.

Comment (NiLuJe) [reply plugins/autodim.koplugin/main.lua:327]: I'm assuming the ramping is fast for that last statement, if that isn't actually the case, then, yeah, update the footer ;).

(Ping @zwim)
